### ISSUE TITLE
tf.data.Dataset.map + tf.numpy_function = lost shape sadness

### ISSUE THREAD

**Issue by SumNeuron** [tensorflow/tensorflow#31373]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): I slightly modified a tf documentation code
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): 16.04 LTS / macOS 14
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: n/a
- TensorFlow installed from (source or binary): 1.14
- TensorFlow version (use command below): conda
- Python version: 3.7
- Bazel version (if compiling from source): n/a
- GCC/Compiler version (if compiling from source): n/a
- CUDA/cuDNN version: colab
- GPU model and memory: colab

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
`tf.numpy_function` works are wrapping numpy / python functions with tf 1.14 in eager and normal mode, but shape is lost

```python
# slight modification from "Load Images with tf.data" https://www.tensorflow.org/tutorials/load_data/images


def preprocess_image(image, resize=[192, 192]):
  image = tf.image.decode_jpeg(image, channels=3)
  image = tf.image.resize(image, [192, 192])
  image /= 255.0  # normalize to [0,1] range

  return image


def load_and_preprocess_image(path, resize=[192, 192]):
  image = tf.read_file(path)
  return preprocess_image(image, resize)

# dataset only contains paths, so wrap whatever value for `resize` in lambda
_load_and_preprocess_image = lambda path: load_and_preprocess_image(path, [192,192])

# we "have" numpy functionality for handling images so wrap in `tf.numpy_function`
tf_load_and_preprocess_image = lambda path: tf.numpy_function(_load_and_preprocess_image, [path], tf.float32)

# map
image_ds2_error_boogaloo = path_ds.map(tf_load_and_preprocess_image, num_parallel_calls=AUTOTUNE)

# no shape
image_ds2_error_boogaloo
# `<DatasetV1Adapter shapes: , types: tf.float32>`

```

**Describe the expected behavior**

shape is not lost

supposedly fixed here #16052 

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

https://colab.research.google.com/drive/1DeGMPxb8cyHm5QLpqJ9-cB2sbF8PC3vA#scrollTo=qGiGN5rl4s2f&line=22&uniqifier=1

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by ravikyram** [comment]
I have tried on colab with TF version 1.14 [gist](https://colab.research.google.com/drive/1w9vKC09IiCJm8X-IQHDGNWxs8WpMPc8J) and nightly versions [gist](https://colab.research.google.com/drive/1Qh1ts8-ijENNopr9yd2pA8_uFBMkxMiQ) was able to reproduce the issue.Thanks!

**Comment by SumNeuron** [comment]
@jvishnuvardhan any updates?

**Comment by deasmhumhna** [comment]
I was dealing with the same problem (in Tensorflow 2.0 beta) and found a workaround of manually setting the `_output_structure` attribute of the `_map_func` (`StructuredFunctionWrapper`) attribute of `MapDataset` (possible thanks to Python's lack of hidden variables) with the proper `NestedStructure` object i.e
```
image_ds._map_func._output_structure = data.experimental.NestedStructure(...)
```
 This suggests a fix of providing an optional structure tuple to `Dataset.map` that can be passed to the  `StructuredFunctionWrapper` to be used in lieu of automatic inference. Not sure how easy that is ; the `StructuredFunctionWrapper` class is fairly opaque to me at first glance.

**Comment by jsimsa** [comment]
@SumNeuron `numpy_function` is opaque to shape inference (as it can contain arbitrary Python code), so the functionality you are asking for is not feasible. Having said that, I don't see why you need to use `numpy_function` for your example. The follow pipeline will work just fine:

```
files = tf.data.Dataset.list_files(...)
images = files.map(lambda f: tf.io.read_file(f))
images = images.map(lambda image: tf.io.decode_jpeg(image))
images = images.map(lambda image: tf.cast(tf.image.resize_images(image, (128, 128)), tf.uint8))
images = images.map(lambda image: tf.cast(image, tf.float32) / 255)
```

@deasmhumhna I would highly discourage your from relying on implementation details of private attributes as these may change between versions and your code will break.

**Comment by SumNeuron** [comment]
@jsimsa the reason for the `numpy_function` inclusion, is because I provided a MWE rather than the actual code. Perhaps one would, inside of `preprocess_image`, do more advanced alterations. See [issue 31235](https://github.com/tensorflow/tensorflow/issues/31235) which includes code copy-pasted from keras's image processing which one might want to use without re writing `numpy` core functions. 

As for "as it can contain arbitrary Python code", true, but in context of applying `numpy` code to `tf.data.Datasets` that doesn't mean it has to exclude attempting shape inference. As stated in the docs:
>Given a python function func, which takes numpy arrays as its arguments and returns numpy arrays as its outputs, wrap this function as an operation in a TensorFlow graph. 
Since it returns arrays as outputs, and numpy arrays have shape this seems like an easy thing to do.

**Comment by ismael-elatifi** [comment]
I had the issue of lost shape because of .map with tf.py_function and I fixed it with the following code (thanks to https://github.com/tensorflow/tensorflow/issues/16052#issue-287936599) :

```python
def set_shapes(img, label, img_shape):
    img.set_shape(img_shape)
    label.set_shape([])
    return img, label

ds = ds.map(lambda img, label: tf.py_function( ... ) )
ds = ds.map(lambda img, label: set_shapes(img, label, img_shape) )
```

**Comment by Dr-Gandalf** [comment]
this shape missing behaviour is a really annoying problem, it defeat the purpose of the py_function, and it should be fixed.

**Comment by dhruvgrover1251** [comment]
To fix this issue, I used the following code :
def map_func(img_name, cap):
 <pre> img_tensor = np.load(img_name) 
 return img_tensor, cap </pre>
dataset = tf.data.Dataset.from_tensor_slices((train_video_paths, train_labels))
def shape(x,y):
<pre>
  x.set_shape((6000,400))
  y.set_shape((48))
  return x,y
</pre>

dataset = dataset.map(lambda item1, item2: tf.numpy_function(map_func, [item1, item2], [tf.float32, tf.int32]),num_parallel_calls=tf.data.experimental.AUTOTUNE)
dataset=dataset.map(shape)

dataset = dataset.shuffle(5).batch(5)
dataset = dataset.prefetch(buffer_size=tf.data.experimental.AUTOTUNE)


But it gives the following error when iterating through the dataset:
InvalidArgumentError: 0-th value returned by pyfunc_7 is double, but expects float
	 [[{{node PyFunc}}]]

**Comment by tobilab** [comment]
> I had the issue of lost shape because of .map with tf.py_function and I fixed it with the following code (thanks to [#16052 (comment)](https://github.com/tensorflow/tensorflow/issues/16052#issue-287936599)) :
> 
> ```python
> def set_shapes(img, label, img_shape):
>     img.set_shape(img_shape)
>     label.set_shape([])
>     return img, label
> 
> img_shape = images[0].shape  # images is a list of numpy.ndarray
> ds = ds.map(lambda img, label: tf.py_function( ... ) )
> ds = ds.map(lambda img, label: set_shapes(img, label, img_shape) )
> ```

Worked for me

**Comment by markub3327** [comment]
@tobilab 

`tf.ensure_shape` works well too and safety.
